pkg/common/plugin/sshpop/sshpop.go

```diff
@@ -0,0 +1,186 @@
+		if err != nil {
+			return nil, Errorf("failed to parse agent svid template: %q", config.AgentPathTemplate)
+		}
+		agentPathTemplate = tmpl
+	}
+	return &Client{
+		cert:              cert,
+		signer:            signer,
+		agentPathTemplate: agentPathTemplate,
+		trustDomain:       trustDomain,
+	}, nil
+}
+
+func stringOrDefault(configValue, defaultValue string) string {
+	if configValue == "" {
+		return defaultValue
+	}
+	return configValue
+}
+
+func getCertAndSignerFromBytes(certBytes, keyBytes []byte) (*ssh.Certificate, ssh.Signer, error) {
+	signer, err := ssh.ParsePrivateKey(keyBytes)
+	if err != nil {
+		return nil, nil, err
+	}
+	pubkey, _, _, _, err := ssh.ParseAuthorizedKey(certBytes)
+	if err != nil {
+		return nil, nil, err
+	}
+	cert, ok := pubkey.(*ssh.Certificate)
+	if !ok {
+		return nil, nil, errors.New("pubkey isn't a certificate")
+	}
+	return cert, signer, nil
+}
+
+func NewServer(trustDomain, configString string) (*Server, error) {
+	if trustDomain == "" {
+		return nil, Errorf("trust_domain global configuration is required")
+	}
+	config := new(ServerConfig)
+	if err := hcl.Decode(config, configString); err != nil {
+		return nil, Errorf("failed to decode configuration: %v", err)
+	}
+	if config.CertAuthorities == nil {
+		return nil, Errorf("missing required config value for \"cert_authorities\"")
+	}
+	certChecker, err := certCheckerFromPubkeys(config.CertAuthorities)
+	if err != nil {
+		return nil, Errorf("failed to create cert checker: %v", err)
+	}
+	agentPathTemplate := DefaultAgentPathTemplate
+	if len(config.AgentPathTemplate) > 0 {
+		tmpl, err := template.New("agent-path").Parse(config.AgentPathTemplate)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse agent svid template: %q", config.AgentPathTemplate)
+		}
+		agentPathTemplate = tmpl
+	}
+	return &Server{
+		certChecker:       certChecker,
+		agentPathTemplate: agentPathTemplate,
+		trustDomain:       trustDomain,
+	}, nil
+}
+
+func certCheckerFromPubkeys(certAuthorities []string) (*ssh.CertChecker, error) {
+	if len(certAuthorities) == 0 {
+		return nil, errors.New("must provide at least one cert authority")
+	}
+	authorities := make(map[string]bool)
+	for _, certAuthority := range certAuthorities {
+		authority, _, _, _, err := ssh.ParseAuthorizedKey([]byte(certAuthority))
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse public key %q: %v", certAuthority, err)
+		}
+		authorities[ssh.FingerprintSHA256(authority)] = true
+	}
+	return &ssh.CertChecker{
+		IsHostAuthority: func(auth ssh.PublicKey, _ string) bool {
+			return authorities[ssh.FingerprintSHA256(auth)]
+		},
+	}, nil
+}
+
+func (c *Client) NewHandshake() *ClientHandshake {
+	return &ClientHandshake{
+		c: c,
+	}
+}
+
+func (s *Server) NewHandshake() *ServerHandshake {
+	return &ServerHandshake{
+		s: s,
+	}
+}
+
+// Errorf is a ssh plugin specific error.
+func Errorf(format string, a ...interface{}) error {
```

Comment:
we've been using the `github.com/zeebo/errs` package for this. errs errors are nice because you can get a stack of the error origination point by printing with `%+v`. this function could be replaced with: 

```
var (
    // unexported package level declaration of the class
    errClass = errs.Class("sshpop")
)

// wherever you are currently calling Errorf
errClass.New("something really %s happened", "bad")
```

Reply:
+1